src/llvmir2hll/graphs/cfg/cfg_traversal.cpp

```diff
@@ -4,6 +4,8 @@
 * @copyright (c) 2017 Avast Software, licensed under the MIT license
```

Comment:
Several tests in our regression test suite in TeamCity have failed (e.g. [these tests](https://retdec-tc.avast.com/viewLog.html?buildId=1280&buildTypeId=Retdec_LinuxBuild&tab=buildLog&_focus=3734)). This seems to be caused by your changes, which indicates that there is probably a bug somewhere. I will try to investigate.

BTW, you can ignore the timeout in the `features.new-static-code-signatures.TestNewARM32Signatures` regression test. This is a generic issue and not related to this PR (see #479).

Reply:
In `master`:
```
retdec-regression-tests-framework $ ./runner.py integration.factorial -r factorial.thumb.gcc.O2.g.elf
Running 1 test case in retdec-regression-tests/integration/factorial for commit 5fb7c9fa...

integration.factorial.Test_2017 (factorial.thumb.gcc.O2.g.elf)                                      [  OK  ]  (0.90s)

SUCCESS (1/1)
```

After the changes in the PR:
```
retdec-regression-tests-framework $ ./runner.py integration.factorial -r factorial.thumb.gcc.O2.g.elf
Running 1 test case in retdec-regression-tests/integration/factorial for commit 09c7bbe0...

integration.factorial.Test_2017 (factorial.thumb.gcc.O2.g.elf)                                      [ FAIL ]  (0.89s)
F
======================================================================
FAIL: test_produce_expected_output (integration.factorial.Test_2017)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "retdec-regression-tests/integration/factorial/test.py", line 8, in test_produce_expected_output
	expected_output='factorial( 5 ) = 120\n'
  File "retdec-regression-tests-framework/regression_tests/tools/decompiler_test.py", line 87, in assert_c_produces_output_when_run
	self.assertEqual(actual_output, expected_output, msg)
AssertionError: 'factorial( 120 ) = 120\n' != 'factorial( 5 ) = 120\n'
- factorial( 120 ) = 120
?            ^^^
+ factorial( 5 ) = 120
?            ^
 : where the input was: 5

----------------------------------------------------------------------
Ran 1 test in 0.077s

FAILED (failures=1)


FAIL (failed 1 out of 1 test)
```

Here is a diff of the generated C codes (the left one is in `master`, the right one is in this PR):

![diff](https://user-images.githubusercontent.com/5675953/52477094-054a5f80-2ba1-11e9-8b3f-f032ee4fa5ef.png)

This might be either a bug in the new CFG traversal or in one of the optimizers that use the CFG traversal (which was triggered by a slightly different traversal order).

You can reproduce the decompilation as follows:
```
wget https://github.com/avast-tl/retdec-regression-tests/raw/master/integration/factorial/2017-11-14/factorial.thumb.gcc.O2.g.elf
python retdec-decompiler-py factorial.thumb.gcc.O2.g.elf
```

Reply:
The difference is caused by `AuxiliaryVariablesOptimizer` (or a combination of `CopyPropagationOptimizer` and `AuxiliaryVariablesOptimizer`). It seems that `AuxiliaryVariablesOptimizer` performs an invalid optimization (`v1` is also used later in the output code, so it cannot be used to store `v5`), but I am still not sure if this is a bug in the optimizer or in the CFG traversal. I will investigate at the beginning of next week.

Reply:
All tests pass now, so I am marking this as resolved.